### ISSUE TITLE
Using CSL for org-export results in 'Wrong type argument' error

### ISSUE THREAD

**Issue by bsag** [bdarcus/citar#527]
**Describe the bug**
I'm using citar within Doom (`:tools biblio` feature). When I try to export an org document using org-export (for example to HTML), export fails with the following error:

```
citeproc-blt-entry-to-csl: Wrong type argument: stringp, nil
```

If I remove the citations and cite_export and print_bibliography keys, the same document exports correctly.

Inserting citations, opening their files and links all works fine. However, if I set `citar-format-reference-function` to `citar-citeproc-format-reference` as in the config below, I get the same error when trying to copy a reference (it asks me for one of my CSL styles first, so it has found the correct directory), so it definitely seems to me to be a problem with citeproc or CSL specifically.

**To Reproduce**
1. Use the config for citar et al. listed below within Doom Emacs, with the biblio feature enabled.
2. Create a minimal org file with some citations inserted.
3. Use org-export-dispatch to choose 'Export as HTML file and open'
4. Get error above and file is not exported to HTML.

**Expected behavior**
Org file is exported to HTML with citations and reference list correctly formatted.

**Emacs version:**
```
Executing 'doom version' with Emacs 28.0.90 at 2022-01-11 17:44:15
  Doom emacs	v21.12.0-alpha   grafted, HEAD -> develop, origin/develop, origin/HEAD 9cba2862 2022-01-10 02:24:21 +0100
  Doom core	v3.0.0-alpha     grafted, HEAD -> develop, origin/develop, origin/HEAD 9cba2862 2022-01-10 02:24:21 +0100
```

**Additional context**
macOS 11.6

Config in `config.el`:

```emacs-lisp
(after! citar
  (setq! citar-bibliography '("~/Documents/bibtex/zotero.bib")
         citar-library-paths '("~/Documents/bibtex/zotero.bib")
         citar-notes-paths '("~/org-roam/literature/")
         citar-citeproc-csl-styles-dir "~/Zotero/styles"
         citar-citeproc-csl-locales-dir "~/Documents/bibtex/locales"
         citar-format-reference-function #'citar-citeproc-format-reference)
  (setq citar-symbols
        `((file ,(all-the-icons-faicon "file-o" :face 'all-the-icons-green :v-adjust -0.1) . " ")
          (note ,(all-the-icons-material "speaker_notes" :face 'all-the-icons-blue :v-adjust -0.3) . " ")
          (link ,(all-the-icons-octicon "link" :face 'all-the-icons-orange :v-adjust 0.01) . " ")))
  (setq citar-symbol-separator "  ")
  ;; use consult-completing-read for enhanced interface
  (advice-add #'completing-read-multiple :override #'consult-completing-read-multiple))

(after! oc
  (setq! org-cite-global-bibliography '("~/Documents/bibtex/zotero.bib")
         org-cite-csl-styles-dir "~/Zotero/styles"))
```

Thanks for any help you can give - I have run out of ideas!

**Comment by bdarcus** [comment]
I doubt this is citar-related ...

> If I remove the citations and cite_export and print_bibliography keys, the same document exports correctly.

What is the content of those lines? Possible there's some subtle problem there?

**Comment by bsag** [comment]
Quite possibly, since I'm not entirely sure what keys I need to just export from org files using citar! I was using the examples on the [announcement page for org-cite](https://blog.tecosaur.com/tmio/2021-07-31-citations.html), so:

```
#+title: Testing citations
#+cite_export: csl apa.csl
...
#+print_bibliography:
```

If I delete the cite_export and print_bibliography keys  but leave the citations in the document and export, the in-text citations are formatted nicely, but there is no reference list/bibliography.

My bib file is BibLaTeX format, exported from Zotero using the Better BibTeX extension. I'm assuming this should work? I could also try using CSL-JSON format, but I think that doesn't include file links which I would miss.

**Comment by bdarcus** [comment]
> #+cite_export: csl apa.csl

With this, org-cite might have trouble finding that file.

Do you have apa.csl in a directory, and have `org-cite-csl-styles-dir` set to it?

Otherwise, use the full path for it and see if that works?

**Comment by bsag** [comment]
I do have that variable set to the directory, but I tried adding the full path for the key like so: `#+cite_export: csl ~/Zotero/styles/apa.csl` but got the same result.

I'm now on my other Mac (exact same version of Doom) and the only difference is that instead of returning with the error reported above, it hangs the org-export dispatch menu until you C-g out. I don't know if it's related there's a message in the Messages buffer like so:

Error during redisplay: (jit-lock-function 1) signaled (wrong-type-argument symbolp (web-mode-markup-indent-offset web-mode-code-indent-offset web-mode-sql-indent-offset web-mode-css-indent-offset))

**Comment by bdarcus** [comment]
Looking at the original error message again, it looks to me citeproc is complaining about either your biblatex file, or one of the entries inside it. It's expected a string, but getting `nil` instead.

In any case, this is definitely not a citar issue.

Any ideas @andras-simonyi?

**Comment by bsag** [comment]
Hmm, odd. Along the way, I have also tried with a much smaller subset of my bib file (which should reduce the chances of formatting problems) but I'll have a go later (probably tomorrow) with a really minimal file with just a couple of entries and see if that sorts the issue. Thanks for taking the time to look at it.

**Comment by andras-simonyi** [comment]
Based on the error message it's most probably an erroneous bibliography entry.

**Comment by bsag** [comment]
That seems likely. However... I've just downloaded the test.bib file in the scripts directory in this repo, and configured that as my bib file. I can view its 85 references fine, but I get exactly the same results (export hanging, copying a reference formatted using CSL hanging) as I do with my own bib file, so I'm not convinced that is the issue. It probably is nothing to do with citar (so feel free to close this issue), but can you suggest what I should try debugging-wise next?

I should note that I found I could use pandoc (with its own version of citeproc) on the my test document, with my own bib file (at the command line, not through Emacs) and it formatted in-text citations and built the reference list fine. So whatever the problem is, pandoc's citeproc isn't affected.

**Comment by bdarcus** [comment]
What's the value for `bibtex-dialect`?

**Comment by andras-simonyi** [comment]
> That seems likely. However... I've just downloaded the test.bib file in the scripts directory in this repo, and configured that as my bib file. I can view its 85 references fine, but I get exactly the same results (export hanging, copying a reference formatted using CSL hanging) as I do with my own bib file, so I'm not convinced that is the issue. 

This is really strange -- I downloaded the same test.bib file and could export _all_ entries using org-cite + apa.csl without any problems. As for debugging the citeproc-el error with your original bibliography, I should and will improve error reporting, but for now, unfortunately, I don't really have a better suggestion than dividing the bib file into two halves, trying both of them separately, then halving the (a) problematic half again etc.

**Comment by bsag** [comment]
How do I check that?

**Comment by bsag** [comment]
@andras-simonyi That's really weird. I'm sure it must be something weird that I've messed up somehow, or something else that I've got installed which is causing a conflict somehow, but I really can't see what it is. I'll try your strategy with my bib file, but if test.bib isn't working for me (and is for you), my problem must be something other than the bib file itself mustn't it?

**Comment by andras-simonyi** [comment]
@bsag  I assumed that you are not seeing the concrete

    citeproc-blt-entry-to-csl: Wrong type argument: stringp, nil

message with the `test.bib` file. If you do then you are right, bisecting your original bibliography doesn't make much sense, at least for now.  Anyhow, I'd first try to make things work with `test.bib` which we can assume to be correct and debug the original bibliography afterwards, if it still doesn't work.

**Comment by bsag** [comment]
Sometimes I see a concrete error, and sometimes the command just hangs and doesn't output an error - the maddening thing is that I can't seem to see a pattern in it. I have also got a similar error of `citeproc-bt--to-csl-date: Wrong type argument: stringp, nil`. Reducing the bib file to a single reference, citing only that reference in the test org file and checking the bib file with a biblatex linter (which returns no errors or warnings), still gets me either a hang or the error.

I'm baffled. Anyway, I'll continue to poke at it. Thanks for your help!

**Comment by andras-simonyi** [comment]
This might be a long shot but I recall seeing strange behavior concerning dates and citeproc when different Emacs versions and their compiled code were mixed up so maybe recompiling things (e.g., citeproc-el) could also be an option.

**Comment by bsag** [comment]
@andras-simonyi Yes!! I've finally sorted it. In case anyone else comes across a similar issue, here's what fixed it.

Your 'long shot' was one part of the solution. I had 3 problems:

- citeproc.el evidently did need to be recompiled as you suggested
- For some reason the CSL files I had downloaded from Zotero were causing issues. The ones I use with pandoc (which should be the same, but who knows...) work.

After fixing those two, I created and configured a very tiny bib file with just two of the references from my bib file. I then cited those two references and it worked.

From there I used `biber --tool -V zotero.bib` to find issues in my full bib file and then went laboriously through my references in Zotero to find and fix the issues identified. Once that was done and Zotero exported the cleaned full bib file, all was finally well. Phew. Thanks again for your help.

**Comment by andras-simonyi** [comment]
@bsag  I'm glad it's working now! and if you encounter any citeproc-el related problem don't hesitate to open at issue at the project repository.

**Comment by bsag** [comment]
@andras-simonyi Will do, but everything is working perfectly together for me now. Between citar, citeproc-el and org-roam, I'm loving the streamlined system for making literature notes and citing references! Thanks to you and @bdarcus for all your hard work!

**Comment by MaxwellMorris** [comment]
Please note that a similar error: 
org-cite-csl-finalizer: Wrong type argument: numberp, nil
Can be resolved by ensuring that you have at least 1 reference in your .org file.

**Comment by AlecVercruysse** [comment]
I'm having the problem @MaxwellMorris mentioned. This is an issue for me since I have the cite_export option set globally: 

`(setq org-cite-export-processors '((t . (csl "~/repos/emacs-config/ieee.csl"))))`

with a global bibliography, so any org file I attempt to export without at least one reference in it gives me `Wrong type argument: numberp, nil` somewhere (varying locations) in `org-cite-csl--generate-latex-preamble`. 

Any suggestions? Can I assume this is a problem with org-cite/my org-cite config? I ask here since this seems to be where the issue is being discussed right now.

**Comment by bdarcus** [comment]
@AlecVercruysse I don't believe, per discussion above, this is a citar issue.

It looks like the functions that are failing are in org?

Do you have any ideas @andras-simonyi?

**Comment by andras-simonyi** [comment]
Thanks for reporting! I've added a simple fix in citeproc-el (commit c6f5cca), hopefully it solves the problem.

**Comment by AlecVercruysse** [comment]
It does, thank you very much!
